[WTF] Shrink sizeof(JSON::Value)

Constellation opened the pull request:
#### 1e2385d6ab1270307f31561a7f9ff2343f7b09a9
<pre>
[WTF] Shrink sizeof(JSON::Value)
<a href="https://bugs.webkit.org/show_bug.cgi?id=259384">https://bugs.webkit.org/show_bug.cgi?id=259384</a>
rdar://112637340

Reviewed by Devin Rousso and Darin Adler.

It makes sizeof(JSON::Value) optimized, and wipe all virtual functions for that.
We would like to use this class more widely, for easy-to-use JSON without any JSC contexts.
One target is many toolings, and for example, JSC bytecode profiler is now using this.

* Source/WTF/wtf/JSONValues.cpp:
(WTF::JSONImpl::Value::visitDerived):
(WTF::JSONImpl::Value::visitDerived const):
(WTF::JSONImpl::Value::operator delete):
(WTF::JSONImpl::Value::asObject):
(WTF::JSONImpl::Value::asObject const):
(WTF::JSONImpl::Value::asArray):
(WTF::JSONImpl::Value::writeJSON const):
(WTF::JSONImpl::Value::writeJSONImpl const):
(WTF::JSONImpl::Value::memoryCost const):
(WTF::JSONImpl::Value::memoryCostImpl const):
(WTF::JSONImpl::ObjectBase::memoryCostImpl const):
(WTF::JSONImpl::ObjectBase::writeJSONImpl const):
(WTF::JSONImpl::ArrayBase::writeJSONImpl const):
(WTF::JSONImpl::ArrayBase::memoryCostImpl const):
(WTF::JSONImpl::ObjectBase::~ObjectBase): Deleted.
(WTF::JSONImpl::ObjectBase::asObject): Deleted.
(WTF::JSONImpl::ObjectBase::asObject const): Deleted.
(WTF::JSONImpl::ObjectBase::memoryCost const): Deleted.
(WTF::JSONImpl::ObjectBase::writeJSON const): Deleted.
(WTF::JSONImpl::ArrayBase::~ArrayBase): Deleted.
(WTF::JSONImpl::ArrayBase::asArray): Deleted.
(WTF::JSONImpl::ArrayBase::writeJSON const): Deleted.
(WTF::JSONImpl::ArrayBase::memoryCost const): Deleted.
* Source/WTF/wtf/JSONValues.h:

Canonical link: <a href="https://commits.webkit.org/266204@main">https://commits.webkit.org/266204@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8dfc5615f3ef9619318d4a7cc1439788fdecf104

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13201 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13514 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13848 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14939 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12568 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13269 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16022 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13541 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15262 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13368 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14041 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11159 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15497 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11329 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11913 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18977 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/11230 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12404 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12080 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15299 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/12492 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12569 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/10437 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/13249 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11842 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3483 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16163 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/13630 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1497 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12412 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3279 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/JSONValues.cpp:
```diff
@@ -492,16 +525,55 @@ RefPtr<Value> Value::asValue()
+    case Type::Array:
+        return nullptr;
+    case Type::Object:
+        static_assert(sizeof(Object) == sizeof(ObjectBase), "cannot cast");
```
No need for this string. Not sure it adds anything.

Constellation replied:
Dropped.

darinadler commented on Source/WTF/wtf/JSONValues.cpp:
```diff
@@ -492,16 +525,55 @@ RefPtr<Value> Value::asValue()
 
 RefPtr<Object> Value::asObject()
 {
```
Now that these functions are so super-simple, maybe they should be inlined?

Constellation replied:
Yeah, changed.

darinadler commented on Source/WTF/wtf/JSONValues.cpp:
```diff
@@ -807,7 +872,7 @@ Ref<Array> Array::create()
-size_t ArrayBase::memoryCost() const
+size_t ArrayBase::memoryCostImpl() const
 {
     size_t memoryCost = Value::memoryCost();
```
This needs to be changed to `memoryCostImpl` to avoid infinite recursion.

Constellation replied:
Oops, right. Fixed.

darinadler commented on Source/WTF/wtf/JSONValues.cpp:
```diff
@@ -646,36 +725,30 @@ void Value::writeJSON(StringBuilder& output) const
-size_t ObjectBase::memoryCost() const
+size_t ObjectBase::memoryCostImpl() const
 {
     size_t memoryCost = Value::memoryCost();
```
This needs to be changed to `memoryCostImpl` to avoid infinite recursion.

Constellation replied:
Fixed.

darinadler commented on Source/WTF/wtf/JSONValues.cpp:
```diff
@@ -646,36 +725,30 @@ void Value::writeJSON(StringBuilder& output) const
+
+size_t Value::memoryCostImpl() const
 {
     size_t memoryCost = sizeof(this);
```
This looks wrong. Shouldn’t it be `sizeof(*this)`?

Constellation replied:
Looks right. Changed.

dcrousso commented on Source/WTF/wtf/JSONValues.cpp:
```diff
@@ -460,6 +461,38 @@ RefPtr<JSON::Value> buildValue(const CodeUnit* start, const CodeUnit* end, const
+
+void Value::operator delete(Value* value, std::destroying_delete_t)
+{
+    value->visitDerived([](auto& value) {
```
NIT: could we maybe use a different name for the lambda parameter `value` so that it's not the same as the outer function parameter `value`?  maybe `derived`?

ditto below as well :)

Constellation replied:
Changed it to `derived`.

dcrousso commented on Source/WTF/wtf/JSONValues.h:
```diff
@@ -156,6 +157,11 @@ class WTF_EXPORT_PRIVATE Value : public RefCounted<Value> {
     }
 
 private:
+    template<typename Visitor> constexpr decltype(auto) visitDerived(Visitor&&);
+    template<typename Visitor> constexpr decltype(auto) visitDerived(Visitor&&) const;
```
NIT: do these actually need to be part of the class?  i.e. could they instead be `static` functions inside the `.cpp` that accept a `Value&` parameter?

Constellation replied:
I personally like this (having it in JSON::Value itself) since how to dispatch this is largely depending on how we construct the class hierarchy of Value.

darinadler commented on Source/WTF/wtf/JSONValues.cpp:
```diff
@@ -646,38 +666,32 @@ void Value::writeJSON(StringBuilder& output) const
+size_t ObjectBase::memoryCostImpl() const
 {
-    size_t memoryCost = Value::memoryCost();
+    size_t memoryCost = Value::memoryCostImpl();
```
I thought about this more and I think we would be better off with `sizeof(*this)` here. Because calling `Value::memoryCostImpl` returns the wrong size, the size just of `Value`, and does a pointless check if it’s a string and we know it’s not. I don’t think we need a call through to the base class for future maintainability either.

Also I think that ideally we should include the memory used by the map itself, not just the memory used for the pointer to the hash table and for the keys and values, but also the memory used by the hash table. I’m honestly not sure how good our `memoryCost` functions needs to be to make garbage collection do the right thing. I suppose estimates are good enough. We could add a `memoryCost` function to HashMap but I suppose it would tricky to decide what to name it to make clear what it does and does not include.

Constellation replied:
Sounds good. changed to `sizeof(*this)`.

dcrousso commented on Source/WTF/wtf/JSONValues.cpp:
```diff
@@ -460,49 +461,61 @@ RefPtr<JSON::Value> buildValue(const CodeUnit* start, const CodeUnit* end, const
+    case Type::Double:
+    case Type::Integer:
+    case Type::String:
+        return std::invoke(std::forward<Visitor>(visitor), static_cast<Value&>(*this));
```
NIT: is the `static_cast` actually needed here since it's a member function of `Value`?  or is the thinking that we'd wanna `static_cast` anyways if we're operating on a subclass?

Constellation replied:
This is not necessary because this is the same class, but I think this is good-to-have to align to the other things (so, we can always use this form even if we add yet another type here).

Constellation committed:
[WTF] Shrink sizeof(JSON::Value)
https://bugs.webkit.org/show_bug.cgi?id=259384
rdar://112637340

Reviewed by Devin Rousso and Darin Adler.

It makes sizeof(JSON::Value) optimized, and wipe all virtual functions for that.
We would like to use this class more widely, for easy-to-use JSON without any JSC contexts.
One target is many toolings, and for example, JSC bytecode profiler is now using this.

* Source/WTF/wtf/JSONValues.cpp:
(WTF::JSONImpl::Value::visitDerived):
(WTF::JSONImpl::Value::visitDerived const):
(WTF::JSONImpl::Value::operator delete):
(WTF::JSONImpl::Value::asObject):
(WTF::JSONImpl::Value::asObject const):
(WTF::JSONImpl::Value::asArray):
(WTF::JSONImpl::Value::writeJSON const):
(WTF::JSONImpl::Value::writeJSONImpl const):
(WTF::JSONImpl::Value::memoryCost const):
(WTF::JSONImpl::Value::memoryCostImpl const):
(WTF::JSONImpl::ObjectBase::memoryCostImpl const):
(WTF::JSONImpl::ObjectBase::writeJSONImpl const):
(WTF::JSONImpl::ArrayBase::writeJSONImpl const):
(WTF::JSONImpl::ArrayBase::memoryCostImpl const):
(WTF::JSONImpl::ObjectBase::~ObjectBase): Deleted.
(WTF::JSONImpl::ObjectBase::asObject): Deleted.
(WTF::JSONImpl::ObjectBase::asObject const): Deleted.
(WTF::JSONImpl::ObjectBase::memoryCost const): Deleted.
(WTF::JSONImpl::ObjectBase::writeJSON const): Deleted.
(WTF::JSONImpl::ArrayBase::~ArrayBase): Deleted.
(WTF::JSONImpl::ArrayBase::asArray): Deleted.
(WTF::JSONImpl::ArrayBase::writeJSON const): Deleted.
(WTF::JSONImpl::ArrayBase::memoryCost const): Deleted.
* Source/WTF/wtf/JSONValues.h:

Canonical link: https://commits.webkit.org/266204@main
